scenes/game/ai/ai_resource.gd

```diff
@@ -0,0 +1,178 @@
+## A data structure for AI manipulation. Behaves mostly like a named tuple (qv.
+## Python), in that it's like a dictionary with extra steps.
+##
+## This mainly exists so that there's specific support for object duplication
+## and value-based equality checking. By default, Godot objects cannot be
+## duplicated (and Resources only support duplication for objects whose _init
+## contains no required arguments), and always check equality by reference only.
+##
+## A class that extends AIResource should implement:
+##
+##   * _init
+##     * Takes zero required arguments.
+##     * If all arguments have their default value, the resulting object should
+##       be as close to a blank slate as possible; for example, if there are
+##       _init behaviors that perform a sync-to-original, these should *not*
+##       happen if all values are default.
+##   * copy
+##     * This is probably just a call to copy_impl but also passes in the
+##       identity of the subclass.
+
+class_name AIResource
+extends Resource
+
+# AIResources usually reflect data coming from a particular source. We retain a
+# reference to that source since we'll sometimes want to sync our values to it.
+# Copies of an AIResource instead point to the original.
+var original
+
+const IGNORE_PROPERTIES = {  # Don't duplicate these properties in copy()
+	# Godot built-ins
+	'RefCounted': 1, 'script': 1, 'Built-in script': 1, 'Resource': 1,
```

Comment:
why is this a dictionary and not an array?

Reply:
O(1) lookup time on keys, though it doesn't really matter for such low values of n.